### PR TITLE
Add UT correctness and address comments for previous symbolic shape PR

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: 
- add correctness check for UT
- address all other comments from https://github.com/microsoft/onnxruntime/pull/6892
- fix graph rebuilt every time when incoming shape is different

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -633,9 +635,10 @@ def test_bert_inputs_with_dynamic_shape():
         x, y, z = _get_bert_for_sequence_classification_sample_data_with_random_shapes('cuda')
 
         outputs = model(x, y, None, None, None, None, z)
-        s = outputs[0]
-        s.backward()
+        loss = outputs[0]
+        loss.backward()
 
+        assert loss is not None
```
Comment (ytaous) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:641]: The output for loss varies a lot even with same inputs. Tried even 1e-1 for atol param, the diff could be out of the range sometimes. It seems that there's something else random in the bert model, thus can't apply torch.allclose() or _test_helpers.assert_gradients_match_and_reset_gradient(). THerefore I am adding simple check just to ensure dynamic shape inputs would work.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -607,23 +607,25 @@ def run_step(model1, model2, model3, x1, x2):
...
-        s.backward()
-
-        assert x.grad is not None 
-        for param in model.parameters():
-            assert param.grad is not None
-            param.grad = None
+        pt_p = run_step(pt_model, x)
+        ort_p = run_step(ort_model, x)
 
+        # assert torch.allclose(ort_p, pt_p)    # TODO: this assert is failing, need to investigate!!
```
Comment (ytaous) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:627]: encountering the same issue on assert, thus disable for now

Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: works for atol=1e-6, testing atol=1e-7

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:641]: are you comparing loss to a fixed value? there is some dropout op in the bert model. If you wish to compare it to a fixed value, you would need to fix a seed on the random value. try torch.manual_seed()

But I would be interested to compare the loss produced by a pytorch run. Is that possible ?

File: orttraining/orttraining/core/graph/gradient_builder_base.cc [hunk 3]
```diff
@@ -59,35 +59,32 @@ void ComputeBroadcastBackwardAxes(
...
       auto A_dim = A_dims[i].dim_param();
       auto B_dim = B_dims[j].dim_value();
 
       if (B_dim != 1) {
-        LOGS_DEFAULT(WARNING) << "Gradient building error for node " << node_name << ": symbolic broadcasting requires the B_dimension to be 1. " <<
-                  "A_dims:" << ToString(A_dims) << ", B_dims:" << ToString(B_dims);  
-        --i;
-        --j;            
-        continue;
+        LOGS_DEFAULT(WARNING) << "Gradient building for node " << node_name << ": symbolic broadcasting expects the B_dimension to be 1. " <<
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.cc:71]: I have another idea for improving this. 
We can keep a mapping of the assumptions we made during gradient graph building,  for symbolic_shape->symbolic_shape and symbolic_shape->concrete_shape. 
We can validate the assumptions when we got the read input data.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc [hunk 4]
```diff
@@ -59,35 +59,32 @@ void ComputeBroadcastBackwardAxes(
...
+                  "A_dims:" << ToString(A_dims) << ", B_dims:" << ToString(B_dims) <<   
+                  " This is a relaxing case, and the kernel might run into problem later if A_dims and B_dims turns out not broadcastable.";   
       }
-      if (B_axes) {
+      if (B_axes && B_dim == 1) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.cc:75]: should this be a "else-if" case?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -607,23 +607,25 @@ def run_step(model1, model2, model3, x1, x2):
...
-        s.backward()
-
-        assert x.grad is not None 
-        for param in model.parameters():
-            assert param.grad is not None
-            param.grad = None
+        pt_p = run_step(pt_model, x)
+        ort_p = run_step(ort_model, x)
 
+        assert torch.allclose(ort_p, pt_p, atol=1e-6)    # relaxing tolerance, 1e-7 or less would fail
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:627]: 1e-6 is ok. 
see tol setting in assert_gradients_match_and_reset_gradient.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:641]: No, I was comparing loss between ortmodule and pytorch run, let me try torch.manual_seed() and see if that helps, thx.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc (hunk 4 above)
Comment (ytaous) [reply orttraining/orttraining/core/graph/gradient_builder_base.cc:75]: logic is the same, but I will make the suggested change

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 5 above)
Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:627]: cool, would this also help on other UTs with TODOs?

File: orttraining/orttraining/core/graph/gradient_builder_base.cc (hunk 3 above)
Comment (ytaous) [reply orttraining/orttraining/core/graph/gradient_builder_base.cc:71]: sure, will sync with you on the idea when starting

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:641]: Still unable to produce tow closed loss values, checking around ...

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 6]
```diff
@@ -607,38 +609,50 @@ def run_step(model1, model2, model3, x1, x2):
...
         x, y, z = _get_bert_for_sequence_classification_sample_data_with_random_shapes('cuda')
 
-        outputs = model(x, y, None, None, None, None, z)
-        s = outputs[0]
-        s.backward()
+        pt_p = run_step(pt_model, x, y, z)
+        ort_p = run_step(ort_model, x, y, z)
+
+        assert torch.allclose(ort_p, pt_p, atol=1e-02)      # TODO: this assert is failing with smaller tolerance, need to investigate!!
+        # _test_helpers.assert_gradients_match_and_reset_gradient(ort_model, pt_model)  #TODO - enable this check after the investigation
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:654]: does setting atol=1e-02 working for this assert?

Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:654]: No, but 1e-1 will, so I leave it disabled for now.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 7]
```diff
@@ -118,7 +118,7 @@ def _forward(self, *inputs, **kwargs):
                 self._input_names_require_grad = input_names_require_grad
                 self._initialize_module_gradient_graph_builder()
 
-            if self._current_input_shape is None or self._current_input_shape != new_input_shape:
+            if self._current_input_shape is None:
                 self._current_input_shape = new_input_shape
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:122]: `self._current_input_shape` will become obsolete and lead to confusion. If we don't care about the exact shape, this should be a boolean

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]
